File: leetcode_vs/234.回文链表.go
package leetcode_vs

/*
 * @lc app=leetcode.cn id=234 lang=golang
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	if head == nil {
		return true
	}
	fast := head
	slow := head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	p := slow.Next
	slow.Next = nil
	for p != nil {
		r := p.Next
		p.Next = slow.Next
		slow.Next = p
		p = r
	}

	fast = slow.Next
	for head != nil && fast != nil && fast.Val == head.Val {
		head = head.Next
		fast = fast.Next

	}
	if fast != nil {
		return false
	}

	return true

}

// @lc code=end

File: 234.回文链表.java
/*
 * @lc app=leetcode.cn id=234 lang=java
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null) {
            return true;
        }
        ListNode slow, fast;
        slow = fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode p, r;

        p = slow.next;
        slow.next = null;
        while (p != null) {
            r = p.next;
            p.next = slow.next;
            slow.next = p;
            p = r;
        }
        fast = slow.next;
        while (head != null && fast != null && head.val == fast.val) {
            head = head.next;
            fast = fast.next;
        }
        if (fast != null) {
            return false;
        }

        return true;
    }
}
// @lc code=end
